homeassistant/components/matter/lock.py

```diff
@@ -0,0 +1,131 @@
+from typing import Any
+
+from chip.clusters import Objects as clusters
+
+from homeassistant.components.lock import LockEntity, LockEntityDescription
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import LOGGER
+from .entity import MatterEntity
+from .helpers import get_matter
+from .models import MatterDiscoverySchema
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Matter lock from Config Entry."""
+    matter = get_matter(hass)
+    matter.register_platform_handler(Platform.LOCK, async_add_entities)
+
+
+class MatterLock(MatterEntity, LockEntity):
+    """Representation of a Matter lock."""
+
+    features: int | None = None
+
+    @property
+    def supports_door_position_sensor(self) -> bool:
+        """Return True if the lock supports door position sensor."""
+        if self.features is None:
+            return False
+
+        return bool(self.features & DoorLockFeature.kDoorPositionSensor)
+
+    async def send_device_command(
+        self,
+        command: clusters.ClusterCommand,
+        timed_request_timeout_ms: int = 1000,
+    ) -> None:
+        """Send a command to the device."""
+        await self.matter_client.send_device_command(
+            node_id=self._endpoint.node.node_id,
+            endpoint_id=self._endpoint.endpoint_id,
+            command=command,
+            timed_request_timeout_ms=timed_request_timeout_ms,
+        )
+
+    async def async_lock(self, **kwargs: Any) -> None:
+        """Lock the lock with pin if needed."""
+
+        await self.send_device_command(command=clusters.DoorLock.Commands.LockDoor())
+
+    async def async_unlock(self, **kwargs: Any) -> None:
+        """Unlock the lock with pin if needed."""
+
+        await self.send_device_command(command=clusters.DoorLock.Commands.UnlockDoor())
+
+    @callback
+    def _update_from_device(self) -> None:
+        """Update the entity from the device."""
+
+        if self.features is None:
+            self.features = int(
+                self.get_matter_attribute_value(clusters.DoorLock.Attributes.FeatureMap)
+            )
+
+        lock_state = self.get_matter_attribute_value(
+            clusters.DoorLock.Attributes.LockState
+        )
+
+        assert lock_state is not None
+
+        LOGGER.debug("Lock state: %s for %s", lock_state, self.entity_id)
+
+        if lock_state is clusters.DoorLock.Enums.DlLockState.kLocked:
+            self._attr_is_locked = True
+        elif lock_state is clusters.DoorLock.Enums.DlLockState.kUnlocked:
+            self._attr_is_locked = False
+        else:
+            self._attr_is_locked = None
+            LOGGER.warning("Unknown lock state: %s for %s", lock_state, self.entity_id)
```

Comment:
Can this happen?

Reply:
Yeah sorta, it can happen when the device is transitioning from a locked to an unlocked state or vise-versa I just don't think there is a nice way to handle that state. other than ignoring it since it doesn't tell you what the current state is so setting the is_unlocking/is_locking attributes is not really possible and won't be 100% accurate

Reply:
What other device states are there?

We do know what the current entity state is, if that helps?

Reply:
The only other state is not fully locked, ill see if I can maybe use it to set the is_unlocking and is_locking attributes